File: src/main/java/no/nav/bidrag/beregn/saertilskudd/rest/mapper/BPAndelSaertilskuddCoreMapper.java
package no.nav.bidrag.beregn.saertilskudd.rest.mapper;


import java.util.ArrayList;
import java.util.Map;
import no.nav.bidrag.beregn.bpsandelsaertilskudd.dto.BeregnBPsAndelSaertilskuddGrunnlagCore;
import no.nav.bidrag.beregn.bpsandelsaertilskudd.dto.InntektPeriodeCore;
import no.nav.bidrag.beregn.bpsandelsaertilskudd.dto.NettoSaertilskuddPeriodeCore;
import no.nav.bidrag.beregn.felles.enums.SjablonTallNavn;
import no.nav.bidrag.beregn.saertilskudd.rest.consumer.SjablonListe;
import no.nav.bidrag.beregn.saertilskudd.rest.dto.http.BMInntekt;
import no.nav.bidrag.beregn.saertilskudd.rest.dto.http.BPInntekt;
import no.nav.bidrag.beregn.saertilskudd.rest.dto.http.BeregnTotalSaertilskuddGrunnlag;
import no.nav.bidrag.beregn.saertilskudd.rest.dto.http.Grunnlag;
import no.nav.bidrag.beregn.saertilskudd.rest.dto.http.InntektRolle;
import no.nav.bidrag.beregn.saertilskudd.rest.dto.http.NettoSaertilskudd;
import no.nav.bidrag.beregn.saertilskudd.rest.dto.http.Rolle;
import no.nav.bidrag.beregn.saertilskudd.rest.dto.http.SBInntekt;

public class BPAndelSaertilskuddCoreMapper extends CoreMapper {

  public BeregnBPsAndelSaertilskuddGrunnlagCore mapBPsAndelSaertilskuddGrunnlagTilCore(
      BeregnTotalSaertilskuddGrunnlag beregnTotalSaertilskuddGrunnlag,
      Map<String, SjablonTallNavn> sjablontallMap, SjablonListe sjablonListe) {
    var nettoSaertilskuddPeriodeListe = new ArrayList<NettoSaertilskuddPeriodeCore>();
    var inntektBPPeriodeListe = new ArrayList<InntektPeriodeCore>();
    var inntektBMPeriodeListe = new ArrayList<InntektPeriodeCore>();
    var inntektBBPeriodeListe = new ArrayList<InntektPeriodeCore>();

    // Løper gjennom alle grunnlagene og identifiserer de som skal mappes til bidragsevne core
    for (Grunnlag grunnlag : beregnTotalSaertilskuddGrunnlag.getGrunnlagListe()) {
      switch (grunnlag.getType()) {
        case INNTEKT -> {
          InntektRolle inntektRolle = grunnlagTilObjekt(grunnlag, InntektRolle.class);
          Rolle rolle = inntektRolle.getRolle();
          if (rolle.equals(rolle.BP)) {
            BPInntekt bpInntekt = grunnlagTilObjekt(grunnlag, BPInntekt.class);
            inntektBPPeriodeListe.add(bpInntekt.tilBPsAndelSaertilskuddCore(grunnlag.getReferanse()));
          } else if (rolle.equals(rolle.BM)) {
            BMInntekt bmInntekt = grunnlagTilObjekt(grunnlag, BMInntekt.class);
            inntektBMPeriodeListe.add(bmInntekt.tilCore(grunnlag.getReferanse()));
          } else if (rolle.equals(rolle.SB)) {
            SBInntekt sbInntekt = grunnlagTilObjekt(grunnlag, SBInntekt.class);
            inntektBBPeriodeListe.add(sbInntekt.tilCore(grunnlag.getReferanse()));
          }
        }
        case NETTO_SAERTILSKUDD -> {
          NettoSaertilskudd nettoSaertilskudd = grunnlagTilObjekt(grunnlag, NettoSaertilskudd.class);
          nettoSaertilskuddPeriodeListe.add(nettoSaertilskudd.tilCore(grunnlag.getReferanse()));
        }
      }
    }

    // Hent aktuelle sjabloner
    var sjablonPeriodeCoreListe = mapSjablonSjablontall(sjablonListe.getSjablonSjablontallResponse(), BP_ANDEL_SAERTILSKUDD,
        beregnTotalSaertilskuddGrunnlag, sjablontallMap);

    return new BeregnBPsAndelSaertilskuddGrunnlagCore(beregnTotalSaertilskuddGrunnlag.getBeregnDatoFra(),
        beregnTotalSaertilskuddGrunnlag.getBeregnDatoTil(), nettoSaertilskuddPeriodeListe, inntektBPPeriodeListe, inntektBMPeriodeListe,
        inntektBBPeriodeListe, sjablonPeriodeCoreListe);
  }
}

File: src/main/kotlin/no/nav/bidrag/beregn/saertilskudd/rest/mapper/BPAndelSaertilskuddCoreMapper.kt
package no.nav.bidrag.beregn.saertilskudd.rest.mapper

import no.nav.bidrag.beregn.bpsandelsaertilskudd.dto.BeregnBPsAndelSaertilskuddGrunnlagCore
import no.nav.bidrag.beregn.bpsandelsaertilskudd.dto.InntektPeriodeCore
import no.nav.bidrag.beregn.bpsandelsaertilskudd.dto.NettoSaertilskuddPeriodeCore
import no.nav.bidrag.beregn.felles.enums.SjablonTallNavn
import no.nav.bidrag.beregn.saertilskudd.rest.consumer.SjablonListe
import no.nav.bidrag.beregn.saertilskudd.rest.dto.http.BMInntekt
import no.nav.bidrag.beregn.saertilskudd.rest.dto.http.BPInntekt
import no.nav.bidrag.beregn.saertilskudd.rest.dto.http.BeregnTotalSaertilskuddGrunnlag
import no.nav.bidrag.beregn.saertilskudd.rest.dto.http.GrunnlagType
import no.nav.bidrag.beregn.saertilskudd.rest.dto.http.InntektRolle
import no.nav.bidrag.beregn.saertilskudd.rest.dto.http.NettoSaertilskudd
import no.nav.bidrag.beregn.saertilskudd.rest.dto.http.Rolle
import no.nav.bidrag.beregn.saertilskudd.rest.dto.http.SBInntekt

object BPAndelSaertilskuddCoreMapper : CoreMapper() {
    fun mapBPsAndelSaertilskuddGrunnlagTilCore(
        beregnTotalSaertilskuddGrunnlag: BeregnTotalSaertilskuddGrunnlag, sjablontallMap: Map<String, SjablonTallNavn>, sjablonListe: SjablonListe
    ): BeregnBPsAndelSaertilskuddGrunnlagCore {
        val nettoSaertilskuddPeriodeListe = ArrayList<NettoSaertilskuddPeriodeCore>()
        val inntektBPPeriodeListe = ArrayList<InntektPeriodeCore>()
        val inntektBMPeriodeListe = ArrayList<InntektPeriodeCore>()
        val inntektBBPeriodeListe = ArrayList<InntektPeriodeCore>()

        // Løper gjennom alle grunnlagene og identifiserer de som skal mappes til bidragsevne core
        for (grunnlag in beregnTotalSaertilskuddGrunnlag.grunnlagListe!!) {
            when (grunnlag.type) {
                GrunnlagType.INNTEKT -> {
                    val (rolle) = grunnlagTilObjekt(grunnlag, InntektRolle::class.java)
                    if (rolle == Rolle.BP) {
                        val bpInntekt = grunnlagTilObjekt(grunnlag, BPInntekt::class.java)
                        inntektBPPeriodeListe.add(bpInntekt.tilBPsAndelSaertilskuddCore(grunnlag.referanse!!))
                    } else if (rolle == Rolle.BM) {
                        val bmInntekt = grunnlagTilObjekt(grunnlag, BMInntekt::class.java)
                        inntektBMPeriodeListe.add(bmInntekt.tilCore(grunnlag.referanse!!))
                    } else if (rolle == Rolle.SB) {
                        val sbInntekt = grunnlagTilObjekt(grunnlag, SBInntekt::class.java)
                        inntektBBPeriodeListe.add(sbInntekt.tilCore(grunnlag.referanse!!))
                    }
                }

                GrunnlagType.NETTO_SAERTILSKUDD -> {
                    val nettoSaertilskudd = grunnlagTilObjekt(grunnlag, NettoSaertilskudd::class.java)
                    nettoSaertilskuddPeriodeListe.add(nettoSaertilskudd.tilCore(grunnlag.referanse!!))
                }

                else -> {}
            }
        }

        // Hent aktuelle sjabloner
        val sjablonPeriodeCoreListe = mapSjablonSjablontall(
            sjablonListe.sjablonSjablontallResponse, BP_ANDEL_SAERTILSKUDD,
            beregnTotalSaertilskuddGrunnlag, sjablontallMap
        )
        return BeregnBPsAndelSaertilskuddGrunnlagCore(
            beregnTotalSaertilskuddGrunnlag.beregnDatoFra!!,
            beregnTotalSaertilskuddGrunnlag.beregnDatoTil!!, nettoSaertilskuddPeriodeListe, inntektBPPeriodeListe, inntektBMPeriodeListe,
            inntektBBPeriodeListe, sjablonPeriodeCoreListe
        )
    }
}
